### PR TITLE
[BEAM-7008] adding UTF8 String coder to Java SDK ModelCoders

### PR DESCRIPTION
**Pull request by ihji**
make StringUtf8Coder standard in Java SDK.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 1]
```diff
@@ -433,6 +433,18 @@ def decode(self, encoded):
     return encoded
 
 
+class StrUtf8CoderImpl(StreamCoderImpl):
+  """For internal use only; no backwards-compatibility guarantees."""
+  def encode_to_stream(self, value, out, nested):
+    byte_value = value.encode('utf-8')
+    out.write_var_int64(len(byte_value))
```
Comment (robertwb) [inline sdks/python/apache_beam/coders/coder_impl.py]:
```
if nested:
    out.write_var_int64(len(byte_value))
```

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 2]
```diff
@@ -433,6 +433,18 @@ def decode(self, encoded):
...
+    out.write(byte_value)
+
+  def decode_from_stream(self, in_stream, nested):
+    byte_length = in_stream.read_var_int64()
```
Comment (robertwb) [inline sdks/python/apache_beam/coders/coder_impl.py]: Similarly. Otherwise read to the end of the stream. This can be done with `stream.read_all(nested)`.

**Comment by robertwb** [comment]
FYI, the current implementation uses https://github.com/apache/beam/blob/release-2.12.0/sdks/python/apache_beam/coders/coder_impl.py#L199 to prefix with the length iff the context is nested.

**Comment by ihji** [comment]
retest this please

File: sdks/python/apache_beam/coders/coders.py [hunk 3]
```diff
@@ -379,6 +359,29 @@ def _create_impl(self):
     raise NotImplementedError
 
 
+class StrUtf8Coder(FastCoder):
```
Comment (robertwb) [inline sdks/python/apache_beam/coders/coders.py]: Why was this moved?

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 4]
```diff
@@ -433,6 +433,22 @@ def decode(self, encoded):
...
+    if nested:
+      out.write_var_int64(len(byte_value))
+    out.write(byte_value)
+
+  def decode_from_stream(self, in_stream, nested):
+    if nested:
+      byte_length = in_stream.read_var_int64()
+      return in_stream.read(byte_length).decode('utf-8')
+    else:
+      return in_stream.read_all(nested).decode('utf-8')
```
Comment (robertwb) [inline sdks/python/apache_beam/coders/coder_impl.py]: This line can handle both the nested and unnested case.

File: sdks/python/apache_beam/coders/coder_impl.py [hunk 5]
```diff
@@ -433,6 +433,22 @@ def decode(self, encoded):
     return encoded
 
 
+class StrUtf8CoderImpl(StreamCoderImpl):
+  """For internal use only; no backwards-compatibility guarantees."""
+  def encode_to_stream(self, value, out, nested):
+    byte_value = value.encode('utf-8')
+    if nested:
+      out.write_var_int64(len(byte_value))
+    out.write(byte_value)
```
Comment (robertwb) [inline sdks/python/apache_beam/coders/coder_impl.py]: This can be the one-liner

`out.write(value.encode('utf-8'), nested)`.

File: sdks/python/apache_beam/coders/coder_impl.py (hunk 4 above)
Comment (ihji) [reply sdks/python/apache_beam/coders/coder_impl.py]: Aha. You're right. I didn't know that length-prefixing is already handled in `stream` code. Thanks for the pointer.

File: sdks/python/apache_beam/coders/coders.py (hunk 3 above)
Comment (ihji) [reply sdks/python/apache_beam/coders/coders.py]: It depends on `FastCoder`.

**Comment by ihji** [comment]
run java precommit

**Comment by ihji** [comment]
run java precommit

**Comment by ihji** [comment]
run java_examples_dataflow precommit

**Comment by ihji** [comment]
run portable_python precommit

**Comment by ihji** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by ihji** [comment]
@robertwb turned out that python strutf8coder modification was not necessary for #8174.

**Comment by ihji** [comment]
run java precommit

**Comment by ihji** [comment]
run python precommit

**Comment by ihji** [comment]
run java precommit

**Comment by ihji** [comment]
Run java precommit

**Comment by robertwb** [comment]
Let me know when this is ready for me to take another look.

**Comment by ihji** [comment]
run java precommit

**Comment by ihji** [comment]
@robertwb PTAL. Thanks!

**Comment by mxm** [comment]
Thanks @ihji!

**Comment by angoenka** [comment]
Run Java PortabilityApi PostCommit
